### ISSUE TITLE
Any plan to release trained models?

### ISSUE THREAD

**Issue by yifeiwang77** [JonasGeiping/fullbatchtraining#2]
RT

**Comment by JonasGeiping** [comment]
Hi, yeah I'm planning to release some models in tandem with the final version the paper. This might still be a few more months away though. 
Which models in particular would be most helpful to you? This would be good to know for a model release later on, but also I might have some checkpoints for your specific case that I can send out already.

**Comment by yifeiwang77** [comment]
Thanks for your reply! It might be helpful to provide the checkpoints that correspond to the main setting in the reported results, such as those in Table 2. I have tried to run your ```train.sh``` script while finding it hard to tell which setting corresponds to the ones in the paper. Maybe more annotations are needed:)

Btw, while simply running ```python train_with_gradient_descent.py name=fbaug_1 hyp=fb1```, there are cases that I ran into degenerated results (20% acc) because the training collapses when it is almost finished. Any comment on that and how to take care of this?

**Comment by JonasGeiping** [comment]
Thanks for the feedback. Yeah, the results from table 2 are definitely on the list for a checkpoint release. Regarding `train.sh`, it's good to know that this needs more annotations. The runs in lines 6-12, for example correspond directly to rows in table 2, with `fbaug_1` being the  "baseline FB" and so forth. I've added annotations.

The checkpoint question is a very pratical one that we could not spend a lot of time on in the paper. During training spiking instabilities occur as you noted (and cf. Fig.2 in the paper) during which both train and test accuracy is terrible. The model always recovers after the spike, but it cannot end on it. 
If training finishes during a spike there are two options: Either continue training for more steps or revert to the last iteration before the spike. In our paper we wanted to keep the maximum budget fixed, so we chose the second option. 

Reverting to the last iteration can be done in multiple ways. In the arxiv version we report the validation accuracy at the iteration with minimal training loss in case of a spike, but I've iterated on this since then and the simplest rule that I have so far is to simply always choose the model checkpoint from the last iteration where the full training loss is smaller than 0.1. Another rule that performs similarly would be to take the last checkpoint where the gradient norm is smaller than 0.25. 
Really anything that chooses a checkpoint from before the spike based on training statistics (and without using the validation set!) is fair game here, and the spike is very noticeable in these metrics. This selection is not implemented in the current checkpointing system though, mainly because actually returning these checkpoints before the spike instead of only their statistics was not a priority so far.


Also, I have uploaded a few checkpoints to the repository now. They can be listed via ` torch.hub.list("jonasgeiping/fullbatchtraining")`. I hope to extend this in the future as more reruns for the final paper version finish on our cluster.
